Set pick nested

FredericHeem opened the pull request.

richytong commented on set.js:
```diff
@@ -0,0 +1,65 @@
+  let nested = result
+  while(nested != null && ++index < pathLength){
+    const pathKey = pathArray[index]
+    if(index === (pathLength - 1)){
```
correct me if I'm wrong but == is slightly more performant than ===

richytong commented on set.js:
```diff
@@ -0,0 +1,65 @@
+ *   path string|Array<string|number>,
+ *   value (value=>any)|any
+ *
+ * set(path, value) -> setter obj=>obj
```
setter object=>object

richytong commented on pick.js:
```diff
@@ -16,20 +16,27 @@
   while (++keysIndex < keysLength) {
     const key = keys[keysIndex],
-      value = source[key]
+      value = get(key)(source)
```
we should use the internal getByPath here

richytong reviewed (changes requested):
Let's move the set implementation into a new function _internal/setByPath, similar to _internal/getByPath. Then we should have the top level api pick and set depend on the _internal setByPath.

FredericHeem commented:
Set is useful for public use, lodash is providing it too.
Regarding code formatting, let's automate it with a command such as "npm
run format"


On Wed, 7 Jul 2021, 18:10 Richard Tong, ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> Let's move the set implementation into a new function _internal/setByPath,
> similar to _internal/getByPath. Then we should have the top level api pick
> and set depend on the _internal setByPath.
> ------------------------------
>
> In set.js
> <https://github.com/a-synchronous/rubico/pull/187#discussion_r665482464>:
>
> > +const get = require("./get")
> +
> +/**
> + * @name set
> + *
> + * @synopsis
> + * ```coffeescript [specscript]
> + * var value any,
> + *   path string|Array<string|number>,
> + *   value (value=>any)|any
> + *
> + * set(path, value) -> setter obj=>obj
> + * ```
> + *
> + * @description
> + * Create a setter that set a property on a value denoted by path.
>
> Create a setter that sets a property on an object denoted by path.
> ------------------------------
>
> In set.js
> <https://github.com/a-synchronous/rubico/pull/187#discussion_r665484650>:
>
> > + * @description
> + * Create a setter that set a property on a value denoted by path.
> + *
> + * ```javascript [playground]
> + * console.log(
> + *   set('a.b', '1')({ a: { c : 2 }),
> + * ) // { a : { b: 1, c: 2 }}
> +  * console.log(
> + *   set('a[0].b.c', 4)({ 'a': [{ 'b': { 'c': 3 } }] }),
> + * ) // { 'a': [{ 'b': { 'c': 4 } }] }
> +*   console.log(set(['a', 'b'], '1')({ a: { c : 2 })),
> + * ) // { a: { b: 1, c: 2 }}
> + * ```
> + */
> +
> +const toPathArray = path => {
>
> we should use the _internal propertyPathToArray function in place of
> toPathArray.
> ------------------------------
>
> In set.js
> <https://github.com/a-synchronous/rubico/pull/187#discussion_r665493972>:
>
> > + * ) // { a : { b: 1, c: 2 }}
> +  * console.log(
> + *   set('a[0].b.c', 4)({ 'a': [{ 'b': { 'c': 3 } }] }),
> + * ) // { 'a': [{ 'b': { 'c': 4 } }] }
> +*   console.log(set(['a', 'b'], '1')({ a: { c : 2 })),
> + * ) // { a: { b: 1, c: 2 }}
> + * ```
> + */
> +
> +const toPathArray = path => {
> +  if(isString(path)) return path.split(".")
> +  if(Array.isArray(path)) return path
> +  throw Error(`path should be a string or an array`)
> +}
> +const set = (path, value) => function setter(obj) {
> +  if((obj == null) || !isObject(obj)){
>
> if statements should have spaces before and after the opening and closing
> parens
>
> core functions should not use x functions internally - use _internal
> functions instead. I believe the internal isObject is the same one as the
> x/isObject. Also we do not need to check if obj == null here - isObject
> will return false if value == null.
> ------------------------------
>
> In set.js
> <https://github.com/a-synchronous/rubico/pull/187#discussion_r665494268>:
>
> > +const toPathArray = path => {
> +  if(isString(path)) return path.split(".")
> +  if(Array.isArray(path)) return path
> +  throw Error(`path should be a string or an array`)
> +}
> +const set = (path, value) => function setter(obj) {
> +  if((obj == null) || !isObject(obj)){
> +    return obj
> +  }
> +  let index = -1
> +  const pathArray = toPathArray(path)
> +  const pathLength = pathArray.length
> +
> +  const result = {...obj}
> +  let nested = result
> +  while(nested != null && ++index < pathLength){
>
> while statements should have spaces before and after opening and closing
> parens as well.
> ------------------------------
>
> In set.js
> <https://github.com/a-synchronous/rubico/pull/187#discussion_r665503301>:
>
> > + */
> +
> +const toPathArray = path => {
> +  if(isString(path)) return path.split(".")
> +  if(Array.isArray(path)) return path
> +  throw Error(`path should be a string or an array`)
> +}
> +const set = (path, value) => function setter(obj) {
> +  if((obj == null) || !isObject(obj)){
> +    return obj
> +  }
> +  let index = -1
> +  const pathArray = toPathArray(path)
> +  const pathLength = pathArray.length
> +
> +  const result = {...obj}
>
> object literals should be spaced like { ...obj }
> ------------------------------
>
> In test.js
> <https://github.com/a-synchronous/rubico/pull/187#discussion_r665505629>:
>
> > @@ -3495,7 +3495,7 @@ flatMap(
>      })
>    })
>
> -  describe('get', () => {
> +  describe('set', () => {
>
> this should still be get
> ------------------------------
>
> In set.js
> <https://github.com/a-synchronous/rubico/pull/187#discussion_r665510940>:
>
> > +  if(Array.isArray(path)) return path
> +  throw Error(`path should be a string or an array`)
> +}
> +const set = (path, value) => function setter(obj) {
> +  if((obj == null) || !isObject(obj)){
> +    return obj
> +  }
> +  let index = -1
> +  const pathArray = toPathArray(path)
> +  const pathLength = pathArray.length
> +
> +  const result = {...obj}
> +  let nested = result
> +  while(nested != null && ++index < pathLength){
> +    const pathKey = pathArray[index]
> +    if(index === (pathLength - 1)){
>
> correct me if I'm wrong but == is slightly more performant than ===
> ------------------------------
>
> In set.js
> <https://github.com/a-synchronous/rubico/pull/187#discussion_r665511825>:
>
> > @@ -0,0 +1,65 @@
> +const isObject = require('./x/isObject')
> +const isString = require('./x/isString')
> +const get = require("./get")
> +
> +/**
> + * @name set
> + *
> + * @synopsis
> + * ```coffeescript [specscript]
> + * var value any,
> + *   path string|Array<string|number>,
> + *   value (value=>any)|any
> + *
> + * set(path, value) -> setter obj=>obj
>
> setter object=>object
> ------------------------------
>
> In pick.js
> <https://github.com/a-synchronous/rubico/pull/187#discussion_r665512526>:
>
> >    let keysIndex = -1
>    while (++keysIndex < keysLength) {
>      const key = keys[keysIndex],
> -      value = source[key]
> +      value = get(key)(source)
>
> we should use the internal getByPath here
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/a-synchronous/rubico/pull/187#pullrequestreview-701154507>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA7NMSIYQ7HZYUTL2P6FOBDTWR37HANCNFSM476OGU7Q>
> .
>

richytong commented:
yes, I like set too. I think it's a fine addition, let's just make it use setByPath internally. get uses getByPath internally, it just performs some null checks and calls getByPath.

FredericHeem commented on set.js:
````diff
@@ -0,0 +1,65 @@
+ * ```
+ */
+
+const toPathArray = path => {
````
I tried with propertyPathToArray but I don't think it should accept an object for this use case

richytong replied:
supplying an object for the path is not common i think. I would rather have consistent behavior with propertyPathToArray.

FredericHeem replied:
when a path is an object, it must throw an error. The path is only valid when it is either a string or an array.

FredericHeem commented on set.js:
```diff
@@ -0,0 +1,65 @@
+  throw Error(`path should be a string or an array`)
+}
+const set = (path, value) => function setter(obj) {
+  if((obj == null) || !isObject(obj)){
```
what are the reasons behind not using public function internally?

richytong commented on pick.js:
```diff
@@ -16,20 +16,27 @@
  */
+
+const getByPath = require("./_internal/getByPath")
+const set = require("./set")
```
this might be what is causing the build step to hang.. maybe. Let's use setByPath here instead of set. It is typically more performant to supply all arguments at once in a single function call as opposed to multiple calls to a higher order function.

FredericHeem replied:
pick is now using setByPath, `node distributor.js` is still hanging

richytong reviewed (approved):
LGTM
